File: DBMS_Lab_Project/DataStructures/HashTableChaining.cpp
#include "HashTableChaining.h"
#include <stdexcept>
#include <sstream>
#include "../BinaryUtils.h"

size_t HashTableChaining::manualHash(const std::string& key) const {
    size_t hash = 5381;
    for (char c : key) {
        hash = ((hash << 5) + hash) + c;
    }
    return hash % capacity;
}

HashTableChaining::HashTableChaining(size_t cap) : capacity(cap) {
    table.resize(capacity);
}

void HashTableChaining::put(const std::string& key, const std::string& value) {
    size_t index = manualHash(key);
    for (auto& pair : table[index]) {
        if (pair.first == key) {
            pair.second = value;
            return;
        }
    }
    table[index].push_back({key, value});
}

std::string HashTableChaining::get(const std::string& key) const {
    size_t index = manualHash(key);
    for (const auto& pair : table[index]) {
        if (pair.first == key) {
            return pair.second;
        }
    }
    throw std::out_of_range("Key not found");
}

void HashTableChaining::remove(const std::string& key) {
    size_t index = manualHash(key);
    auto& chain = table[index];
    for (auto it = chain.begin(); it != chain.end(); ++it) {
        if (it->first == key) {
            chain.erase(it);
            return;
        }
    }
}



std::string HashTableChaining::serialize() const {
    std::stringstream ss;
    size_t total = 0;
    for(const auto& chain : table) total += chain.size();
    writeSize(ss, total);

    for (const auto& chain : table) {
        for (const auto& pair : chain) {
            writeString(ss, pair.first);
            writeString(ss, pair.second);
        }
    }
    return ss.str();
}

void HashTableChaining::deserialize(const std::string& str) {
    table.clear();
    table.resize(capacity);
    if(str.empty()) return;
    std::stringstream ss(str);
    size_t count;
    readSize(ss, count);

    for(size_t i = 0; i < count; ++i) {
        std::string key, value;
        readString(ss, key);
        readString(ss, value);
        put(key, value);
    }
}

File: DBMS_Lab_Project_Go/internal/datastructures/hash_table_chaining.go
package datastructures

import (
	"bytes"
	"errors"
)

type KV struct {
	Key   string
	Value string
}

type HashTableChaining struct {
	table    [][]KV
	capacity int
}

func NewHashTableChaining(cap int) *HashTableChaining {
	if cap <= 0 {
		cap = 16
	}
	return &HashTableChaining{
		table:    make([][]KV, cap),
		capacity: cap,
	}
}

func (ht *HashTableChaining) manualHash(key string) int {
	var hash uint64 = 5381
	for _, c := range key {
		hash = ((hash << 5) + hash) + uint64(c)
	}
	return int(hash % uint64(ht.capacity))
}

func (ht *HashTableChaining) Put(key, value string) {
	idx := ht.manualHash(key)
	for i, kv := range ht.table[idx] {
		if kv.Key == key {
			ht.table[idx][i].Value = value
			return
		}
	}
	ht.table[idx] = append(ht.table[idx], KV{Key: key, Value: value})
}

func (ht *HashTableChaining) Get(key string) (string, error) {
	idx := ht.manualHash(key)
	for _, kv := range ht.table[idx] {
		if kv.Key == key {
			return kv.Value, nil
		}
	}
	return "", errors.New("Key not found")
}

func (ht *HashTableChaining) Remove(key string) {
	idx := ht.manualHash(key)
	for i, kv := range ht.table[idx] {
		if kv.Key == key {
			// Remove element
			ht.table[idx] = append(ht.table[idx][:i], ht.table[idx][i+1:]...)
			return
		}
	}
}

func (ht *HashTableChaining) Serialize() string {
	var buf bytes.Buffer
	total := 0
	for _, chain := range ht.table {
		total += len(chain)
	}
	WriteSize(&buf, total)
	for _, chain := range ht.table {
		for _, kv := range chain {
			WriteString(&buf, kv.Key)
			WriteString(&buf, kv.Value)
		}
	}
	return buf.String()
}

func (ht *HashTableChaining) Deserialize(str string) {
	ht.table = make([][]KV, ht.capacity)
	if str == "" {
		return
	}
	buf := bytes.NewBufferString(str)
	count, _ := ReadSize(buf)
	for i := 0; i < count; i++ {
		k, _ := ReadString(buf)
		v, _ := ReadString(buf)
		ht.Put(k, v)
	}
}
